### PR TITLE
added support and bot for output feeds to an AMQP topic exchange

### PR DESCRIPTION
**Pull request by sch3m4**

### CODE REVIEW

**Comment by aaronkaplan** [comment]
Hi @sch3m4 ! 
Thank you very much for this contribution. I reviewed the code briefly and added some comments of what i noticed. Hope it helps. 
My main question is: can we test the output bot somehow automatically (we run automatic unit tests with Travis-CI).

**Comment by sch3m4** [comment]
Hi Aaron! I've used this gist to get the output: https://gist.github.com/sch3m4/ef466d273334c269b643 I've used autopep8 to format the source code following the recomendations. I did a simple commit to trigger the compilation with Travis-CI (updated the CHANGELOG.md) and it passed: https://travis-ci.org/sch3m4/intelmq/builds/83974410

**Comment by sebix** [comment]
Your branch cannot be merged automatically, please rebase it on upstream master (don't merge!).

**Comment by aaronkaplan** [comment]
On Oct 6, 2015, at 10:50 PM, Chema García <notifications@github.com> wrote:

> Hi Aaron! I've used this gist to get the output: https://gist.github.com/sch3m4/ef466d273334c269b643 I've used autopep8 to format the source code following the recomendations. I did a simple commit to trigger the compilation with Travis-CI (updated the CHANGELOG.md) and it passed: https://travis-ci.org/sch3m4/intelmq/builds/83974410
> 

Okay, thanks!

But actually I was simply wondering how (in theory) we could automatically unit-test an output bot such as the AMQP bot.
As far as I see , there is currently now test code for it. Which prompted the question "how is it even possible to test the AMQP output bot?"

Don't get me wrong - the sky will not fall from above if there is no unit testing code for it.
But I guess it's an interesting question.

File: REQUIREMENTS [hunk 1]
```diff
@@ -17,3 +17,4 @@ six>=1.7
 python-termstyle>=0.1.10
 unicodecsv>=0.9.4
 xmpppy>=0.5.0rc1
+pika>=0.10.0
```
Comment (sebix) [inline REQUIREMENTS]: The lines in this file are sorted, also the REQUIREMENTS2

File: intelmq/bots/outputs/amqptopic/output.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        for i in range(0, self._retries):
+            try:
+                self._connection = pika.BlockingConnection(
+                    pika.URLParameters(self._url))
+                break
+            except pika.exceptions.AMQPConnectionError:
+                self.logger.info(
+                    'Could not connect to server. Retrying... (%d/%d)' % (i + 1, self._retries))
+                time.sleep(self._delay)
+            except Exception, e:
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: This is the deprecated syntax, please use `Exception as exc` (but you don't need the variable, see next comment).

File: intelmq/bots/outputs/amqptopic/output.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                self.logger.exception('Unknown exception: %s', e)
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: You may use the `traceback` module and it's functions `print_exc` or `format_exc` to give the error including traceback information.

File: intelmq/bots/outputs/amqptopic/output.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+        # verify | set connection
+        if None in (self._connection, self._channel):
+            self.connect_server()
+
+        properties = pika.BasicProperties(
+            content_type=self.CONTENT_TYPE, delivery_mode=self.DELIVERY_MODE)
+
+        try:
+            if not self._channel.basic_publish(exchange=self._exchange, routing_key=self._routing_key, body=event.to_json(), properties=properties, mandatory=True):
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: This line is too long, just break the parameters:

                if not self._channel.basic_publish(exchange=self._exchange,
                                                   routing_key=self._routing_key,
                                                   body=event.to_json(),
                                                   properties=properties,
                                                   mandatory=True):

File: intelmq/bots/outputs/amqptopic/output.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            if not self._channel.basic_publish(exchange=self._exchange,
+                                               routing_key=self._routing_key,
+                                               body=event.to_json(),
+                                               properties=properties,
+                                               mandatory=True):
+                self.logger.error("Message sent but not confirmed")
+            self._errors = 0
+        except Exception:
+            self.logger.error("Error publishing the message")
+            traceback.print_exc()
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: Sorry, I totally forgot, that a `logger.exception(msg)` method exists. It's the same as the error function, but adds the traceback automatically.

Comment (sch3m4) [inline intelmq/bots/outputs/amqptopic/output.py]: no problem, I'll use logger.exception for simplicity

**Comment by sch3m4** [comment]
It appears to be an error outside the pull request, could you verify it?

psycopg2
tests/bots/experts/squelcher/test_expert.py

**Comment by sch3m4** [comment]
branch rebased

**Comment by sebix** [comment]
Ok, I will check the code in detail on Tuesday.

@aaronkaplan and @SYNchroACK The travis check fails because you have removed the psycopg2 dependency form the REQUIREMENTS files. You have 2 possibilities to fix this: Only run the psycopg2 tests if psycopg2 is installed or add the installation command for the package manually to the `.travis.yml` (or anything similar).

This PR can be merged anyway, the travis fail is not @sch3m4's fault, current master fails.

File: intelmq/bots/outputs/amqptopic/output.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    def connect_server(self):
+        self.logger.info("Connecting to %s", self._url)
+        for i in range(0, self._retries):
+            try:
+                self._connection = pika.BlockingConnection(
+                    pika.URLParameters(self._url))
+                break
+            except pika.exceptions.AMQPConnectionError:
+                self.logger.info(
+                    'Could not connect to server. Retrying... (%d/%d)' % (i + 1, self._retries))
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: line too long

File: intelmq/bots/outputs/amqptopic/output.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                time.sleep(self._delay)
+            except Exception:
+                self.logger.exception('Unknown exception')
+
+        # connection error
+        if self._connection is None:
+            self.logger.error('Connection retries exceeded: %d', self._retries)
+            # stop the bot
+            self.stop()
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: No, just raise an error. The bot class handles the bot shutdown itself.

File: intelmq/bots/outputs/amqptopic/output.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+        try:
+            if not self._channel.basic_publish(exchange=self._exchange,
+                                               routing_key=self._routing_key,
+                                               body=event.to_json(),
+                                               properties=properties,
+                                               mandatory=True):
+                self.logger.error("Message sent but not confirmed")
+            self._errors = 0
+        except Exception:
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: Can't this also be handled by the bot class? It does already have an error handling, including a retry counter and it stops when reached.

File: intelmq/bots/outputs/amqptopic/output.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        properties = pika.BasicProperties(
+            content_type=self.CONTENT_TYPE, delivery_mode=self.DELIVERY_MODE)
+
+        try:
+            if not self._channel.basic_publish(exchange=self._exchange,
+                                               routing_key=self._routing_key,
+                                               body=event.to_json(),
+                                               properties=properties,
+                                               mandatory=True):
+                self.logger.error("Message sent but not confirmed")
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: https://github.com/certtools/intelmq/blob/master/docs/Developers-Guide.md#log-messages-format
Log messages must end with a period (IMHO an exclamation mark should also do), also on line 76.

**Comment by sebix** [comment]
Haven't tested the code itself (no idea was AMQP is), but had a look at the code itself, see my comments.

It seems you rebased on master after @aaronkaplan made an unintended merge from another repo into this master. Maybe this can be fixed by doing an interactive rebasing (`git rebase -i`, and a refspec) and delete the superfluous commits.

Also, don't make changes in master, always create a new branch. Have a look at this great summary: https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html

**Comment by sebix** [comment]
I missing docs too.

**Comment by aaronkaplan** [comment]
On Oct 13, 2015, at 5:12 PM, Sebastian <notifications@github.com> wrote:

> Haven't tested the code itself (no idea was AMQP is), but had a look at the code itself, see my comments.
> 
> It seems you rebased on master after @aaronkaplan made an unintended merge from another repo into this master. Maybe this can be fixed by doing an interactive rebasing (git rebase -i, and a refspec) and delete the superfluous commits.
> 

Yeah, Chema, very sorry about the unintentional commit and the confusion it caused.
Maybe you can re-do your PR based on the latest version - in order to keep the commit history all nice.

Best,
a.

**Comment by sebix** [comment]
@sch3m4 Don't worry about the rebasing/merging. If the code is ok, I can get the commits from your repository and merge them myself. And did you checkout from certat/intelmq? There are many commits from this repo.

**Comment by sch3m4** [comment]
@sebix I did checkout from certtools/intelmq, that's why I'm sending the pull request to this repository. Should I take as base the certat/intelmq? Should I send the pull request to certat/intelmq?

**Comment by sebix** [comment]
Ok, I had a look at the network graph. You pulled the code, after @aaronkaplan accidentally merged and pushed the certat/intelmq to certtools/master. I fixed that afterwards, but between these two events you pulled the polluted master.

I will give you instructions how to clean your code when I work on intelmq the next time test your code locally, so your time effort is as low as possible.

**Comment by sebix** [comment]
Please never work on master but always on a separate branch.

    git checkout master
    git branch backup master  # First do a backup: (You can also do a copy of your local copy - including .git files!)
    git remote add upstream git@github.com:certtools/intelmq  # if you don't have it already
    git remote update  # fetch the data
    git rebase upstream/master  # rebase on upstream/master

I got no rebasing conflicts when testing it.

**Comment by pedromreis** [comment]
Will this bot be available?

**Comment by sebix** [comment]
Getting this bot into master requires several tasks:
* Get familiar with AMQP
* Set up a test environment with AMQP
* Test the code
* Fix the code
* [Write tests]

This is not very high on my priority list and my time is very limited. It may seem that I'm working all my time on intelmq, which is by no means the case. It's close to the opposite.

**Comment by pedromreis** [comment]
We use RabbitMQ here, I'll try to test this.

**Comment by sebix** [comment]
Great!

**Comment by pedromreis** [comment]
ok, to be usable (at least by me, in our RabbitMQ)  I think this need some tweeks. For once, the password needs to be a separate parameter so we can use special chars (;!:$ etc). Also, some other parameters need to be declared in the exchange topic, in my case I need Durable=True for example. The default is False. next week I'll try to make a proposal.

**Comment by sebix** [comment]
Closed in favor of #526

**Pull request closed without merging**